include/seastar/core/semaphore.hh

```diff
@@ -185,6 +189,35 @@ private:
         }
     };
     internal::abortable_fifo<entry, expiry_handler> _wait_list;
+
+#ifdef SEASTAR_SEMAPHORE_DEBUG
+    struct used_flag {
+        // set to true from the wait path
+        // prevents the semaphore from being moved or move-reassigned when _used
+        bool _used = false;
+
+        used_flag() = default;
+        used_flag(used_flag&& o) noexcept {
+            assert(!_used && "semaphore cannot be moved after it has been used");
+        }
+        used_flag& operator=(used_flag&& o) noexcept {
+            if (this != &o) {
+                assert(!_used && !o._used && "semaphore cannot be moved after it has been used");
+            }
+            return *this;
+        }
+        void use() noexcept {
+            _used = true;
+        }
+    };
+#else
+    struct used_flag {
+        void use() noexcept {}
+    };
+#endif
+
+    used_flag _used;
```

Comment:
Add [[no_unique_address]] to avoid increasing the semaphore size in non-debug builds.

Reply:
in v5